### DISCUSSION TITLE
How to get file uri

### DISCUSSION THREAD

**Discussion by msaquib17** [eclipse-theia/theia#11891] [Q&A] [answered]
I'm using the below code snippet to get the file uri.
```ts
@inject(FrontendApplication) protected readonly application: FrontendApplication;

constructor(@inject(TYPES.IActionDispatcher) protected actionDispatcher: GLSPActionDispatcher) {
    super();
    this.inputElement = document.createElement("input");
}

protected initializeContents(containerElement: HTMLElement): void {
    // do nothing
}

async show(root: Readonly<SModelRoot>, ...contextElementIds: string[]) {
    super.show(root, ...contextElementIds);
    if (this.isEditorEnabled()) {
        if (this.label) {
             console.log(this)
             if (this.application.shell.activeWidget instanceof DiagramWidget) {
                const diagramWidget = this.application.shell.activeWidget;
                const sourceUri=diagramWidget.uri.toString();
```

Is there any alternative way to get the file uri which does not involve DiagramWidget?

**Comment by msujew** [comment]
Hey @msaquib17,

Widgets which are based on resources usually implement the [`Navigatable`](https://github.com/eclipse-theia/theia/blob/90982d2edbe1e7aac776501897f92ca4c873c576/packages/core/src/browser/navigatable-types.ts#L24-L33) interface. Our own editors do that for example, and I believe the [glsp diagram widget does so as well](https://github.com/eclipse/sprotty-theia/blob/efd153a410267d81af555e70c219e9efa3e3b197/src/theia/diagram-widget.ts#L54). You can then use [`NavigatableWidget.getUri()`](https://github.com/eclipse-theia/theia/blob/90982d2edbe1e7aac776501897f92ca4c873c576/packages/core/src/browser/navigatable-types.ts#L66-L70) to get the uri of the resource that the widget represents.

**Reply by msaquib17** [reply]
But the getUri function takes the widget as the argument. Is there a way we can get the current file uri without using widget?

**Reply by msujew** [reply]
How else is Theia supposed to know which URI you want, if you don't supply the widget? Is there a problem in using `this.application.shell.activeWidget`?

**Reply by msaquib17** [reply]
yes actually this.application.shell.activeWidget becomes undefined and we are looking for a way to get file uri without using this.application.shell.activeWidget

**Reply by vince-fugnitto** [reply]
@msaquib17 do you mind describing your use-case a little more? 
- are you only interested in `DiagramWidgets`?
- do you care about it being the current widget, opened, or active?

**Reply by msaquib17** [reply]
I am not interested in DiagramWidget. I want to get the uri of the current file opened. this.application.shell.activeWidget is undefined and I am looking for an alternative way to get the file uri(eg. C:\Users\win10\mamba\client\file1.mtrule2) @vince-fugnitto @msujew

**Reply by msaquib17** [reply]
The reason I am asking this question is that when I previously used SGraphview for the GLSP graph view in `configureModelElement(context, "graph", GLSPGraph, SGraphView);` the this.application.shell.activeWidget was working fine but when I change the view to GLSPProjectionView, like - `configureModelElement(context, "graph", GLSPGraph, GLSPProjectionView);` the this.application.shell.activeWidget becomes undefined. Hence, I am looking for a way to get the file uri with some other way. Can you please provide some guidance on what is going wrong and how to fix it.

**Reply by vince-fugnitto** [reply]
@msaquib17 unfortunately I'm not an expert in `GLSP` or have any experience with it as it is a completely separate project to the Eclipse Theia framework (you may have luck asking there https://github.com/eclipse-glsp/glsp/discussions).

The `activeWidget` relies on the widget actually preserving and having focus, if that's not what you want you can try something like `currentWidget` or even iterating the list of opened widgets to get the underlying resource you're looking for:

- [`activeWidget` docs](https://eclipse-theia.github.io/theia/docs/next/classes/core.ApplicationShell-3.html#activeWidget)
- [`currentWidget` docs](https://eclipse-theia.github.io/theia/docs/next/classes/core.ApplicationShell-3.html#currentWidget)

**Reply by msaquib17** [reply]
Leaving aside glsp. Is there a way to get the current file path in theia(eg. C:\Users\win10\mamba\client\file1.mtrule2) @vince-fugnitto

**Reply by vince-fugnitto** [reply]
@msaquib17 yes, if the editor is opened then you can get the file path, is that what you mean? You can use the `EditorManager` to get the current editor and then find it's path.

**Reply by msaquib17** [reply]
Yes that's what I mean. Can you please give some more details. Like where is the EditorManager located in @theia? (like import { EditorManager } from "@theia/core..."). Also, to get the file uri, is this the correct syntax? - EditorManager._currentEditor().uri? @vince-fugnitto

**Reply by vince-fugnitto** [reply]
@msaquib17 sure, a simple search should show that `EditorManager` comes from `@theia/editor`:

- [Sources](https://github.com/eclipse-theia/theia/blob/master/packages/editor/src/browser/editor-manager.ts)
- [API Documentation](https://eclipse-theia.github.io/theia/docs/next/classes/editor.EditorManager-1.html)

> is this the correct syntax?

Please take a look at the available types and methods that are provided. The `currentEditor` will return an `EditorWidget` if it exists and from it you can get the underlying resource through `getResourceUri`.

- [API Documentation for `EditorWidget`](https://eclipse-theia.github.io/theia/docs/next/classes/editor.EditorWidget-1.html)

**Reply by msaquib17** [reply]
@vince-fugnitto thanks it worked!!

**Comment by vince-fugnitto** [comment, marked as answer]
Resolution: https://github.com/eclipse-theia/theia/discussions/11891#discussioncomment-4207166.
